File: libpebble3/src/iosMain/kotlin/io/rebble/libpebblecommon/js/JSTimeout.js
globalThis._LibPebbleTimeoutCallbacks = new Map();
globalThis._LibPebbleTriggerTimeout = function (timeoutId) {
    if (globalThis._LibPebbleTimeoutCallbacks.has(timeoutId)) {
        const { callback, args } = globalThis._LibPebbleTimeoutCallbacks.get(timeoutId);
        callback(...args);
        globalThis._LibPebbleTimeoutCallbacks.delete(timeoutId);
    }
}
globalThis._LibPebbleTriggerInterval = function (intervalId) {
    if (globalThis._LibPebbleTimeoutCallbacks.has(intervalId)) {
        const { callback, args } = globalThis._LibPebbleTimeoutCallbacks.get(intervalId);
        callback(...args);
    }
}
globalThis.setTimeout = function (callback, delay, ...args) {
    if (typeof callback !== 'function') {
        throw new TypeError('First argument must be a function');
    }

    const timeoutId = _Timeout.setTimeout(delay);
    globalThis._LibPebbleTimeoutCallbacks.set(timeoutId, { callback, args });
    return timeoutId;
}
globalThis.clearTimeout = function (timeoutId) {
    if (timeoutId === undefined || timeoutId === null) {
        return;
    }
    if (typeof timeoutId !== 'number') {
        throw new TypeError('First argument must be a number');
    }

    if (globalThis._LibPebbleTimeoutCallbacks.has(timeoutId)) {
        globalThis._Timeout.clearTimeout(timeoutId);
        globalThis._LibPebbleTimeoutCallbacks.delete(timeoutId);
    }
}
globalThis.setInterval = function (callback, delay, ...args) {
    if (typeof callback !== 'function') {
        throw new TypeError('First argument must be a function');
    }

    const intervalId = _Timeout.setInterval(delay);
    globalThis._LibPebbleTimeoutCallbacks.set(intervalId, { callback, args });
    return intervalId;
}
globalThis.clearInterval = function (intervalId) {
    if (intervalId === undefined || intervalId === null) {
        return;
    }
    if (typeof intervalId !== 'number') {
        throw new TypeError('First argument must be a number');
    }

    if (globalThis._LibPebbleTimeoutCallbacks.has(intervalId)) {
        globalThis._Timeout.clearInterval(intervalId);
        globalThis._LibPebbleTimeoutCallbacks.delete(intervalId);
    }
}

File: libpebble3/src/iosMain/kotlin/io/rebble/libpebblecommon/js/JSTimeout.kt
package io.rebble.libpebblecommon.js

import co.touchlab.kermit.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import platform.JavaScriptCore.JSContext
import platform.JavaScriptCore.JSManagedValue
import platform.JavaScriptCore.JSValue
import kotlin.time.Duration.Companion.milliseconds

class JSTimeout(private val scope: CoroutineScope, private val evalRaw: (String) -> JSValue?): RegisterableJsInterface {
    private val timeouts = mutableMapOf<Int, Job>()
    private var timeoutIDs = (1..Int.MAX_VALUE).iterator()
    private val jsFunTriggerTimeout: JSManagedValue get() = JSManagedValue(evalRaw("globalThis._LibPebbleTriggerTimeout")!!)
    private val jsFunTriggerInterval: JSManagedValue get() = JSManagedValue(evalRaw("globalThis._LibPebbleTriggerInterval")!!)
    override val interf = mapOf(
        "setTimeout" to this::setTimeout,
        "setInterval" to this::setInterval,
        "clearTimeout" to this::clearTimeout,
        "clearInterval" to this::clearInterval
    )
    override val name = "_Timeout"

    override fun dispatch(method: String, args: List<Any?>) = when (method) {
        "setTimeout" -> setTimeout((args[0] as Number).toDouble())
        "setInterval" -> setInterval((args[0] as Number).toDouble())
        "clearTimeout" -> { clearTimeout((args[0] as Number).toInt()); null }
        "clearInterval" -> { clearInterval((args[0] as Number).toInt()); null }
        else -> error("Unknown method: $method")
    }

    private fun triggerTimeout(id: Int) {
        jsFunTriggerTimeout.value?.callWithArguments(listOf(id.toDouble()))
    }
    private fun triggerInterval(id: Int) {
        jsFunTriggerInterval.value?.callWithArguments(listOf(id.toDouble()))
    }

    fun setTimeout(delay: Double): Double {
        val id = getNextTimeoutID()
        val job = scope.launch {
            delay(delay.milliseconds)
            if (isActive) {
                triggerTimeout(id)
            }
        }
        timeouts[id] = job
        return id.toDouble()
    }

    fun setInterval(delay: Double): Double {
        val id = getNextTimeoutID()
        val job = scope.launch {
            while (isActive) {
                delay(delay.milliseconds)
                if (isActive) {
                    triggerInterval(id)
                }
            }
        }
        timeouts[id] = job
        return id.toDouble()
    }

    fun clearTimeout(id: Int) {
        timeouts.remove(id)?.cancel("Cleared by clearTimeout")
    }

    fun clearInterval(id: Int) {
        timeouts.remove(id)?.cancel("Cleared by clearInterval")
    }

    private fun getNextTimeoutID(): Int {
        return if (timeoutIDs.hasNext()) {
            timeoutIDs.next()
        } else {
            timeoutIDs = (0..Int.MAX_VALUE).iterator()
            timeoutIDs.next()
        }
    }

    override fun close() {
        Logger.d("Closing JSTimeout, cancelling all timeouts")
        scope.cancel("JSTimeout closed")
        timeouts.clear()
    }
}